torchtext/models/roberta/bundler.py

```diff
@@ -97,6 +105,40 @@ def get_model(self, head: Optional[Module] = None, load_weights: bool = True, fr
             model.load_state_dict(state_dict, strict=True)
         return model
 
+    @classmethod
+    def from_config(
+        self,
+        config: RobertaEncoderConf,
+        head: Optional[Module] = None,
+        freeze_encoder: bool = False,
+        checkpoint: Optional[Union[str, Dict[str, torch.Tensor]]] = None,
+        *,
+        dl_kwargs: Dict[str, Any] = None,
+    ) -> RobertaModel:
+        """Class method to intantiate model with user-defined encoder configuration and checkpoint
+
+        Args:
+            config (RobertaEncoderConf): An instance of class RobertaEncoderConf that defined the encoder configuration
+            head (nn.Module, optional): A module to be attached to the encoder to perform specific task
+            freeze_encoder (bool): Indicates whether to freeze the encoder weights
+            checkpoint (str or Dict[str, torch.Tensor], optional): Path to or actual model state_dict. state_dict can have partial weights i.e only for encoder.
+            dl_kwargs (dictionary of keyword arguments): Passed to :func:`torch.hub.load_state_dict_from_url`.
+        """
+        model = _get_model(config, head, freeze_encoder)
+        if checkpoint is not None:
+            if torch.jit.isinstance(checkpoint, Dict[str, torch.Tensor]):
+                state_dict = checkpoint
+            elif isinstance(checkpoint, str):
+                dl_kwargs = {} if dl_kwargs is None else dl_kwargs
+                state_dict = load_state_dict_from_url(checkpoint, **dl_kwargs)
+            else:
+                raise TypeError("checkpoint must be of type `str` or `Dict[str, torch.Tensor]` but got {}".format(type(checkpoint)))
+            if head is not None:
+                model.load_state_dict(state_dict, strict=False)
```

Comment:
I feel like there is an ambiguity in the logic flow here.
Is checkpoint expected to contain parameter for head or does not it matter?

Reply:
Related thread https://github.com/pytorch/text/pull/1424#discussion_r736932480

The use-case I am interested in is:
1) User can optionally start from pre-trained encoder weights provided by the the library (or otherwise)
2) User can optionally start from checkpoint saved by the user
or
3) User can just start from scratch (although in this case Alternative 2 in summary above is also equivalent)

The checkpoint could optionally contain the weights of head (for example to cover point 2 above), or only contain the encoder weights (for example to cover point 1 above). 

Now if head is not None, and we are in scenario 1, then setting strict=True will cause key error. I wonder if there are better ways to deal with this? (Perhaps augment the state_dict with head state_dict if head state_dict is not present and then always set strict=True?) 


BTW, I understand that this is bit unconventional the way we treat the model and check-points in the context of pre-trained models (that typically have one-2-one relationship with model and their checkpoints). I am open to other suggestions to cover the use-case above, but I felt that putting such an API in bundler is potentially a safer choice than creating explicit builder functions to deal with such compositional requirements.